pkg/networkservice/metrics/server.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+	"github.com/sirupsen/logrus"
+
+	"github.com/networkservicemesh/sdk-vppagent/pkg/networkservice/vppagent"
+)
+
+// Server - helper interface to get some various information required for testing.
+type Server interface {
+	// GetMetricsReceived - show how many events we receive from internal vpp agent inerface
+	GetMetricsReceived() int
+	networkservice.NetworkServiceServer
+}
+
+type metricsServer struct {
+	interval        time.Duration
+	connections     map[string]*connectionInfo
+	vppConnections  map[string]string
+	executor        serialize.Executor
+	vppClient       configurator.StatsPollerServiceClient
+	metricsRecieved int
+	Server
+}
+
+type connectionInfo struct {
+	connectionID string
+	index        uint32
+	metrics      map[string]string
+	ifName       string
+}
+
+func (s *metricsServer) GetMetricsReceived() (result int) {
+	<-s.executor.AsyncExec(func() {
+		result = s.metricsRecieved
+	})
+	return
+}
+
+func (s *metricsServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	conf := vppagent.Config(ctx)
+	if conf == nil {
+		return nil, errors.New("VPPAgent config is missing")
+	}
+	ifaces := conf.GetVppConfig().GetInterfaces()
+	if len(ifaces) == 0 {
+		return nil, errors.New("VPPAgent config should contain at least one interface")
+	}
+
+	index := request.GetConnection().GetPath().GetIndex()
+
+	conn, err := next.Server(ctx).Request(ctx, request)
+	if err == nil {
+		<-s.executor.AsyncExec(func() {
+			info := s.connections[conn.GetId()]
+			if info == nil {
+				info = &connectionInfo{
+					connectionID: conn.GetId(),
+					index:        conn.GetPath().GetIndex(),
+					metrics:      map[string]string{},
+					ifName:       ifaces[0].Name,
+				}
+				s.connections[conn.GetId()] = info
+				// Store interface name to store values into
+				s.vppConnections[info.ifName] = info.connectionID
+			}
+			// Update connection with metrics
+			if len(info.metrics) > 0 && len(conn.GetPath().GetPathSegments()) > int(index) {
+				conn.GetPath().GetPathSegments()[index].Metrics = info.metrics
+			}
+		})
+	}
+	return conn, err
+}
+
+func (s *metricsServer) Close(ctx context.Context, conn *networkservice.Connection) (*empty.Empty, error) {
+	s.executor.AsyncExec(func() {
+		info := s.connections[conn.GetId()]
+		if info != nil {
+			delete(s.connections, info.connectionID)
+			delete(s.vppConnections, info.ifName)
+		}
+	})
+	return next.Server(ctx).Close(ctx, conn)
+}
+
+// NewServer creates a new metrics collector instance
+func NewServer(interval time.Duration, vppClient configurator.StatsPollerServiceClient) networkservice.NetworkServiceServer {
+	rv := &metricsServer{
+		connections:    map[string]*connectionInfo{},
+		vppConnections: map[string]string{},
+		interval:       interval,
+		vppClient:      vppClient,
+	}
+	go rv.collect()
```

Comment:
How does this go routine get cleaned up?

Reply:
Good catch, Thanks. fixed.